client/components/with-localized-moment/context.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * External dependencies
+ *
+ * @format
+ */
+
+import React from 'react';
+import { connect } from 'react-redux';
+import moment from 'moment';
+
+/**
+ * Internal dependencies
+ */
+import getCurrentLocaleSlug from 'state/selectors/get-current-locale-slug';
+
+const { Provider, Consumer } = React.createContext( moment );
+
+class MomentProvider extends React.Component {
+	state = {};
+
+	async checkAndLoad() {
+		const { currentLocale } = this.props;
+
+		// is moment set to the current locale?
+		if ( currentLocale === moment.locale() ) {
+			return;
+		}
+
+		if ( currentLocale !== 'en' ) {
+			await import( /* webpackChunkName: "moment-locale-[request]", webpackInclude: /\.js$/ */ `moment/locale/${ currentLocale }` );
+		}
+
+		moment.locale( currentLocale );
+		this.setState( {} );
+	}
+
+	componentDidMount() {
+		this.checkAndLoad();
+	}
+
+	componentDidUpdate() {
+		this.checkAndLoad();
+	}
+
+	render() {
+		return <Provider value={ moment }>{ this.props.children }</Provider>;
```

Comment:
This code was passing the `currentLocale` string as the context value rather than the `moment` object. Fixed in https://github.com/Automattic/wp-calypso/pull/26476/commits/30b3b967ef09ead9c51da5f00025d4bfcdf249bb

Reply:
@jsnajdr I think that was on purpose. The moment instance doesn't change, so it doesn't always cause the re-render necessary when the locale changes.

Reply:
But the component wrapped by `withMoment()` needs to receive a moment instance, not a locale string. Calypso crashed whenever the `moment` prop from `withMoment()` was used. For example, in the "Joined ..." label in `/me/account`

<img width="433" alt="screen shot 2018-10-31 at 15 21 24" src="https://user-images.githubusercontent.com/664258/47794434-bc247d00-dd20-11e8-9ed2-04c48f8c7c50.png">

I'm trying another trick in https://github.com/Automattic/wp-calypso/pull/26476/commits/e16769881266f13e60ccf75444e16bb549e3f62a: set a `{ moment }` object as the value of the context instead of just the constant `moment`. The container object is in local state and updated on locale change. Inspired by https://reactjs.org/docs/context.html#caveats

The only way to break this is to put a pure component between `withMoment` and the site where the `moment` prop is used:
```jsx
class Label extends PureComponent {
  render() {
    return <span>{ this.props.moment( this.props.date ).format() }</span>
  }
}

class LocalizedLabel = withMoment( Label );

<Label date="2018-10-31" />
```

The `LocalizedComponent` itself will rerender on locale change, but the inner `Label` will receive unchanged `moment` and `date` props! Because it's pure, it won't rerender.

Several possible solutions if this turns out to really be a problem:
1. Create a real new instance of `moment` on locale change. Is there any caveat that prevents that?
2. Pass an additional helper `momentLocale` string prop to the child of `withMoment`, with the purpose to get  through `PureComponent` defenses.
3. Instead of relying on context updates, the Moment Provider would be an event emitter and Moment Consumers can subscribe to updates.

Reply:
I don't think 1. is possible, as moment itself is a singleton

I like 2 the best.

Reply:
updated to 2 in https://github.com/Automattic/wp-calypso/pull/26476/commits/04a49d56dda0a358ae792c0f9f6ada09b017ad15